From 7b18488d08056d75da96615319d3a46df98ce0ae Mon Sep 17 00:00:00 2001
From: Rudi <r.ihle@s-t.de>
Date: Fri, 25 Dec 2015 17:12:19 +0100
Subject: [PATCH] video: mxc_edid: Change nearest mode search

The mode match is now performed in up to four passes:
1.) Include modes with exactly the same vmode
2.) Include modes ignoring aspect ratio flags
3.) Include modes with opposite scan but otherwise same vmode
4.) Include modes with opposite scan ignoring aspect ratio flags

The flag FB_VMODE_FRACTIONAL is always ignored.

Signed-off-by: Rudi <r.ihle@s-t.de>
---
 drivers/video/mxc/mxc_edid.c | 63 ++++++++++++++++++++++----------------------
 drivers/video/mxc/mxc_hdmi.c |  4 +--
 include/video/mxc_edid.h     |  2 +-
 3 files changed, 34 insertions(+), 35 deletions(-)

diff --git a/drivers/video/mxc/mxc_edid.c b/drivers/video/mxc/mxc_edid.c
index ef8d261..87bd119 100644
--- a/drivers/video/mxc/mxc_edid.c
+++ b/drivers/video/mxc/mxc_edid.c
@@ -800,52 +800,51 @@ int mxc_edid_read(struct i2c_adapter *adp, unsigned short addr,
 EXPORT_SYMBOL(mxc_edid_read);
 
 const struct fb_videomode *mxc_fb_find_nearest_mode(const struct fb_videomode *mode,
-						    struct list_head *head, bool relax)
+					  struct list_head *head)
 {
 	struct list_head *pos;
 	struct fb_modelist *modelist;
 	struct fb_videomode *cmode;
-	static struct fb_videomode *best;
-	static u32 diff, diff_refresh;
-	u32 mask = relax ? ~FB_VMODE_ASPECT_MASK : ~0;
-	int refresh = mode->refresh;
+	struct fb_videomode *best = NULL;
+	u32 d, diff = ~0, diff_refresh = ~0;
+	int i, refresh = mode->refresh;
+	static const u32 masks[] = {
+		~(FB_VMODE_FRACTIONAL),
+		~(FB_VMODE_FRACTIONAL | FB_VMODE_ASPECT_MASK),
+		~(FB_VMODE_FRACTIONAL | FB_VMODE_INTERLACED),
+		~(FB_VMODE_FRACTIONAL | FB_VMODE_INTERLACED | FB_VMODE_ASPECT_MASK)
+	};
 
 	if ((mode->flag & FB_MODE_IS_FROM_VAR) &&
 	    (mode->vmode & FB_VMODE_FRACTIONAL))
 		refresh--;
 
-	if (!relax) {
-		diff = -1;
-		diff_refresh = -1;
-		best = NULL;
-	}
+	for (i = 0; i < ARRAY_SIZE(masks); i++) {
+		list_for_each(pos, head) {
+			modelist = list_entry(pos, struct fb_modelist, list);
+			cmode = &modelist->mode;
+
+			if ((mode->vmode ^ cmode->vmode) & masks[i])
+					continue;
 
-	list_for_each(pos, head) {
-		u32 d;
-
-		modelist = list_entry(pos, struct fb_modelist, list);
-		cmode = &modelist->mode;
-
-		if ((mode->vmode ^ cmode->vmode) & mask)
-				continue;
-
-		d = abs(cmode->xres - mode->xres) +
-			abs(cmode->yres - mode->yres);
-		if (diff > d) {
-			diff = d;
-			diff_refresh = abs(cmode->refresh - refresh);
-			best = cmode;
-		} else if (diff == d) {
-			d = abs(cmode->refresh - refresh);
-			if (diff_refresh > d) {
-				diff_refresh = d;
+			d = abs(cmode->xres - mode->xres) +
+				abs(cmode->yres - mode->yres);
+			if (diff > d) {
+				diff = d;
+				diff_refresh = abs(cmode->refresh - refresh);
 				best = cmode;
+			} else if (diff == d) {
+				d = abs(cmode->refresh - refresh);
+				if (diff_refresh > d) {
+					diff_refresh = d;
+					best = cmode;
+				}
 			}
 		}
-	}
 
-	if ((!relax && (diff_refresh || diff)) || !best)
-		mxc_fb_find_nearest_mode(mode, head, true);
+		if (diff_refresh == 0 && diff == 0)
+			break;
+	}
 
 	return best;
 }
diff --git a/drivers/video/mxc/mxc_hdmi.c b/drivers/video/mxc/mxc_hdmi.c
index 6156e16..6baf1b6 100644
--- a/drivers/video/mxc/mxc_hdmi.c
+++ b/drivers/video/mxc/mxc_hdmi.c
@@ -2072,7 +2072,7 @@ static void mxc_hdmi_set_mode(struct mxc_hdmi *hdmi, int edid_status)
 
 	fb_var_to_videomode(&m, &var);
 	dump_fb_videomode(&m);
-	mode = mxc_fb_find_nearest_mode(&m, &hdmi->fbi->modelist, false);
+	mode = mxc_fb_find_nearest_mode(&m, &hdmi->fbi->modelist);
 
 	if (mode) {
 		hdmi->fbi->mode = (struct fb_videomode *)mode;
@@ -2378,7 +2378,7 @@ static void mxc_hdmi_setup(struct mxc_hdmi *hdmi, unsigned long event)
 		else
 			memset(&hdmi->prev_virtual, 0, sizeof(hdmi->prev_virtual));
 		if (!list_empty(&hdmi->fbi->modelist)) {
-			edid_mode = mxc_fb_find_nearest_mode(&m, &hdmi->fbi->modelist, false);
+			edid_mode = mxc_fb_find_nearest_mode(&m, &hdmi->fbi->modelist);
 			pr_debug("edid mode ");
 			/* update fbi mode */
 			hdmi->fbi->mode = (struct fb_videomode *)edid_mode;
diff --git a/include/video/mxc_edid.h b/include/video/mxc_edid.h
index 1e89fd2..01a487e 100644
--- a/include/video/mxc_edid.h
+++ b/include/video/mxc_edid.h
@@ -110,6 +110,6 @@ int mxc_edid_read(struct i2c_adapter *adp, unsigned short addr,
 int mxc_edid_parse_ext_blk(unsigned char *edid, struct mxc_edid_cfg *cfg,
 	struct fb_monspecs *specs);
 const struct fb_videomode *mxc_fb_find_nearest_mode(const struct fb_videomode *mode,
-                                                    struct list_head *head, bool relax);
+						struct list_head *head);
 int mxc_fb_mode_is_equal_res(const struct fb_videomode *mode1, const struct fb_videomode *mode2);
 #endif
-- 
1.9.1

